File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.py
class Solution:
    def f(self,n,dp):
        if n<=1:
            return n
        if n==2:
            return 1
            
        if dp[n]!=-1:
            return dp[n]
        
        dp[n]=self.f(n-1,dp)+self.f(n-2,dp)+self.f(n-3,dp)
        return dp[n]
        
    def tribonacci(self, n: int) -> int:
        if n<=1:
            return n
        if n==2:
            return 1

        dp=[-1]*(n+1)
        dp[0]=0
        dp[1]=1
        dp[2]=2
        return self.f(n,dp)
        

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.cpp
class Solution {
public:
    int tribonacci(int n) {
        if(n==0) return 0;
        if(n<=2) return 1;

        int num0=0;
        int num1=1;
        int num2=1;

        for(int i=3;i<=n;i++)
        {
            int temp = num0+num1+num2;
            num0=num1;
            num1=num2;
            num2=temp;
        }

        return num2;
    }
};